File: src/opening_hours.py
from datetime import datetime, time, timedelta
import calendar


class OpeningHours:
    def __init__(self, days, open_time, close_time):
        self.days = [day.lower() for day in days]
        self.open_time = time.fromisoformat(open_time)
        self.close_time = time.fromisoformat(close_time)

    def isOpenOn(self, dt: str):
        local_dt = datetime.fromisoformat(dt.replace("Z", "+00:00"))
        weekday = calendar.day_name[local_dt.weekday()].lower()
        return (
            weekday[:3] in self.days
            and self.open_time <= local_dt.time() <= self.close_time
        )

    def nextOpeningDate(self, dt: str):
        current_dt = datetime.fromisoformat(dt.replace("Z", "+00:00"))
        for i in range(8):
            candidate = current_dt + timedelta(days=i)
            weekday = calendar.day_name[candidate.weekday()].lower()[:3]
            if weekday in self.days:
                opening_datetime = candidate.replace(
                    hour=self.open_time.hour,
                    minute=self.open_time.minute,
                    second=0,
                    microsecond=0,
                )
                if opening_datetime > current_dt:
                    return opening_datetime.isoformat() + "Z"
        return None

File: ts-version/src/OpeningHours.ts
const allDays = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
class OpeningDay {
    constructor(
        public name: string,
        public isOpen: boolean,
    ) {}
}

export class OpeningHours {
    public days: OpeningDay[]
    constructor(
        inputDays: string[],
        public openTime: string,
        public closeTime: string,
    ) {
        this.days = allDays.map(
            (day) => new OpeningDay(day, inputDays.includes(day)),
        )
    }

    withinOpeningHours(datetime: string) {
        const time = new Date(datetime).getHours()
        return (
            time >= parseInt(this.openTime) && time < parseInt(this.closeTime)
        )
    }

    get openDays() {
        return this.days.filter((day) => day.isOpen)
    }

    get openDayNames() {
        return this.days.filter((day) => day.isOpen).map((day) => day.name)
    }

    isOpenOn(date: string) {
        return (
            this.openDayNames.includes(this.dayNameFor(date)) &&
            this.withinOpeningHours(date)
        )
    }

    nextOpeningDate(datetime: string) {
        const sliceAfterDate = this.days.slice(
            this.days
                .map((day) => day.name)
                .indexOf(this.dayNameFor(datetime)) + 1,
        )
        const nextOpenDayIndex =
            [...sliceAfterDate, ...this.days].findIndex((day) => day.isOpen) + 1
        const inputDay = new Date(datetime)
        inputDay.setDate(inputDay.getDate() + nextOpenDayIndex)
        return inputDay.toISOString()
    }

    dayNameFor(date: string) {
        return new Date(date).toLocaleDateString("en-AU", {
            weekday: "short",
        })
    }

    timeFor(date: string) {
        return new Date(date).toLocaleTimeString("en-AU", {
            hour: "2-digit",
            minute: "2-digit",
        })
    }
}
